deoxys/src/deoxys_bc.rs

```diff
@@ -0,0 +1,173 @@
+        }
+
+        subkeys
+    }
+}
+
+impl DeoxysBcInternal for DeoxysBc256 {
+    type SubkeysSize = U15;
+    type TweakKeySize = U32;
+
+    fn generate_subkey(
+        subkey: &mut [u8; 16],
+        tweakey: &GenericArray<u8, Self::TweakKeySize>,
+        index: usize,
+    ) {
+        let rcon = [
+            1,
+            2,
+            4,
+            8,
+            RCON[index],
+            RCON[index],
+            RCON[index],
+            RCON[index],
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+        ];
+
+        for i in 0..16 {
+            subkey[i] = tweakey[i] ^ tweakey[i + 16] ^ rcon[i];
+        }
+    }
+
+    fn shuffle_tweakey(tweakey: &mut GenericArray<u8, Self::TweakKeySize>) {
+        h_substitution(&mut tweakey[16..]);
+        lfsr2(&mut tweakey[..16]);
+        h_substitution(&mut tweakey[..16]);
+    }
+}
+
+impl DeoxysBcType for DeoxysBc256 {
+    type KeySize = U16;
+}
+
+impl DeoxysBcInternal for DeoxysBc384 {
+    type SubkeysSize = U17;
+    type TweakKeySize = U48;
+
+    fn generate_subkey(
+        subkey: &mut [u8; 16],
+        tweakey: &GenericArray<u8, Self::TweakKeySize>,
+        index: usize,
+    ) {
+        let rcon = [
+            1,
+            2,
+            4,
+            8,
+            RCON[index],
+            RCON[index],
+            RCON[index],
+            RCON[index],
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+        ];
+
+        for i in 0..16 {
+            subkey[i] = tweakey[i] ^ tweakey[i + 16] ^ tweakey[i + 32] ^ rcon[i];
+        }
+    }
+
+    fn shuffle_tweakey(tweakey: &mut GenericArray<u8, Self::TweakKeySize>) {
+        h_substitution(&mut tweakey[32..]);
+        lfsr2(&mut tweakey[16..32]);
+        h_substitution(&mut tweakey[16..32]);
+        lfsr3(&mut tweakey[..16]);
+        h_substitution(&mut tweakey[..16]);
+    }
+}
+
+impl DeoxysBcType for DeoxysBc384 {
+    type KeySize = U32;
+}
+
+fn h_substitution(tk: &mut [u8]) {
+    let mut result = [0u8; 16];
+
+    for i in 0..16 {
+        result[i] = tk[H_PERM[i] as usize];
+    }
+
+    tk.copy_from_slice(&result[..16]);
+}
+
+// TODO: This operation is very slow
+// On Deoxys-II-256, shuffle_tweakey(), 
+//   which consists of h_substitution and lfsr2 and lfsr3, takes up 65% of the encryption time
+fn lfsr2(tk: &mut [u8]) {
+    for x in tk {
+        let feedback = (*x >> 5) & 1;
+        *x = x.rotate_left(1);
+        *x ^= feedback;
+    }
+}
+
+// TODO: This operation is very slow
+// On Deoxys-II-256, shuffle_tweakey(), 
+//   which consists of h_substitution and lfsr2 and lfsr3, takes up 65% of the encryption time
+fn lfsr3(tk: &mut [u8]) {
+    for x in tk {
+        let feedback = (*x << 1) & 0x80;
+        *x = x.rotate_right(1);
+        *x ^= feedback;
+    }
+}
```

Comment:
This is the part that's really slow

Reply:
Oof, well my only thought there is a SIMD implementation, but doing a quick search I don't see any crates that make it easy to implement/synthesize that for LFSRs, which is a bit of a shame.

Reply:
While I wouldn't say that's a good solution, I've played around with `unsafe` to interpret the whole input as a u128 and do all bytes at once and it seems to partially help (20% overall performance over the current implementation for the complete encryption/decryption process). Maybe there's way without `unsafe` code to hint the compiler to do it this way? 

```
// TODO: This operation is very slow
// On Deoxys-II-256, shuffle_tweakey(),
//   which consists of h_substitution and lfsr2 and lfsr3, takes up 65% of the encryption time
fn lfsr2(tk: &mut [u8]) {
    unsafe {
        let data = &mut *(tk as *mut [u8] as *mut u128);
        *data = ((*data << 1) & 0xFEFEFEFEFEFEFEFEFEFEFEFEFEFEFEFE) | (((*data >> 7) ^ (*data >> 5)) & 0x01010101010101010101010101010101);
    }
}

// TODO: This operation is very slow
// On Deoxys-II-256, shuffle_tweakey(),
//   which consists of h_substitution and lfsr2 and lfsr3, takes up 65% of the encryption time
fn lfsr3(tk: &mut [u8]) {
    unsafe {
        let data = &mut *(tk as *mut [u8] as *mut u128);
        *data = ((*data >> 1) & 0x7F7F7F7F7F7F7F7F7F7F7F7F7F7F7F7F) | (((*data << 7) ^ (*data << 1)) & 0x80808080808080808080808080808080);
    }
}
```

Reply:
Another thing I just thought is that the actual input key(so the first 128 or 256 bits of the tweakey) and their multiple state for the key schedule can be precomputed as soon as the cipher is created and cached. This would basically mean that the LSFR and the h function would only be called once per round when the cipher is created

Reply:
Now vectorized correctly, closing.